Add a decomposition for torch.put, 2.

pearu opened the pull request:
As in the title. It is an updated copy of https://github.com/pytorch/pytorch/pull/115306 .

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #120179



cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler @amjames @desertfire @chauhang

pearu committed:
Add a decomposition for torch.put, 2.

[ghstack-poisoned]

pearu committed:
Update on "Add a decomposition for torch.put, 2."

As in the title. It is an updated copy of https://github.com/pytorch/pytorch/pull/115306 .




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames

[ghstack-poisoned]

pearu committed:
Update on "Add a decomposition for torch.put, 2."

As in the title. It is an updated copy of https://github.com/pytorch/pytorch/pull/115306 .




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames

[ghstack-poisoned]

pearu committed:
Update on "Add a decomposition for torch.put, 2."

As in the title. It is an updated copy of https://github.com/pytorch/pytorch/pull/115306 .




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames

[ghstack-poisoned]

lezcano commented on torch/_inductor/codegen/cpp.py:
```diff
@@ -1531,8 +1531,6 @@ def load(self, name: str, index: sympy.Expr):
         var = self.args.input(name)
         index = self.rename_indexing(index)
         line = f"{var}[{cexpr_index(index)}]"
-        if V.graph.get_dtype(name) in [torch.float16]:
-            line = f"static_cast<float>({line})"
```
Is this intended?

pearu replied:
yes, otherwise, a value is loaded as `float32` that later is used as a second argument to `atomic_add` leading to compilation failure (because the first argument is of type `float16`).

It is not clear why this mapping is there in the first place.

lezcano replied:
This is to perform the load in low precision, but the computations in a higher precision. There is a similar line in triton
https://github.com/pytorch/pytorch/blob/637cf4a3f2cfdd364005681636ca885bdc4d5887/torch/_inductor/codegen/triton.py#L1919-L1920

I think that the proper fix to your issue is somewhere else, like doing a downcast in the `store` or so.

pearu replied:
Of course, in the case of `put` op, there is no reason to do computations in a higher precision when accumulate is false. Ideally, operations ought to take care of using proper precision when performing computations. Converting to higher precision in `load` for just-in-case does not make sense, unless it is known that computations with float16/bfloat16 are not supported on the device in the first place.

pearu replied:
The original code is now restored with an additional casting in `store` op.

lezcano commented on torch/_decomp/decompositions.py:
```diff
@@ -4497,6 +4497,20 @@ def take(self, index):
+        flattened, [index], source.reshape(index.shape), accumulate
+    )
+    output = flattened.reshape(self.shape)
+    if stride != output.stride():
+        output = prims.copy_strided(output, stride)
```
I'm not sure about this part. I get the feeling that this may force an unwanted copy if it can't be fused with the next operation? What happens if you perform a put into an operation that cannot be fused into, like an `mm` or something that has a fallback?

pearu replied:
When using `put` op as input to `mm` and input strides do not match output strides, there will be an additional triton kernel call of copying the put result prior calling `mm`.

This `if` block is now removed as the function is moved to `torch/_inductor/decompositons.py`

lezcano reviewed (commented):
As discussed internally, I think this should be an inductor-only decomp due to the striding issue.

pearu committed:
Update on "Add a decomposition for torch.put, 2."

As in the title. It is an updated copy of https://github.com/pytorch/pytorch/pull/115306 .




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames

[ghstack-poisoned]

lezcano commented on test/onnx/test_fx_op_consistency.py:
```diff
@@ -1019,6 +1019,15 @@ def skip_torchlib_forward_compatibility(
         "pca_lowrank",
         reason=onnx_test_common.reason_dynamo_does_not_support("wrapper_set_seed"),
     ),
+    xfail(
+        "put",
+        dtypes=onnx_test_common.BOOL_TYPES,
+        reason=onnx_test_common.reason_onnx_script_does_not_support("index_put", "bool"),
+    ),
+    xfail(
+        "put",
+        reason=onnx_test_common.reason_dynamo_does_not_support("fixme: indices element out of data bound"),
+    ),
```
These, can go now, and you should also remove `put` and `put.out` from `has_decomposition`, I believe.

pearu replied:
Removed.

lezcano commented on torch/_decomp/__init__.py:
```diff
@@ -372,6 +372,8 @@ def core_aten_decompositions() -> Dict[torch._ops.OperatorBase, Callable]:
             aten.norm,
             aten.ones,
             aten.ones_like,
+            aten.put,
```
this one should also go

pearu replied:
Removed.

pearu commented on torch/_decomp/decompositions.py:
```diff
@@ -4529,7 +4515,6 @@ def put(self, index, source, accumulate=False):
 register_inplace(aten.__ixor__, aten.__xor__)
 register_inplace(aten.leaky_relu_, aten.leaky_relu)
 register_inplace(aten.logit_, aten.logit)
-register_inplace(aten.put_, aten.put)
```
Fixed.

pearu replied:
Leaving it in leads to recursion error.

lezcano replied:
can you paste the traceback? Does this happen when you put it as part of the core decomp?

pearu replied:
Sure:

<details>

```
$ pytest -sv test/inductor/test_torchinductor_opinfo.py -k test_comprehensive_put -x
=========================================================================================== test session starts ============================================================================================
platform linux -- Python 3.9.1, pytest-7.1.2, pluggy-0.13.1 -- /home/pearu/miniconda3/envs/pytorch-cuda-dev/bin/python
cachedir: .pytest_cache
hypothesis profile 'dev' -> database=None, max_examples=10, suppress_health_check=[HealthCheck.too_slow]
rootdir: /home/pearu/git/pytorch/pytorch-csc, configfile: pytest.ini
plugins: pytorch-0.0.0, xdoctest-1.1.0, hypothesis-6.0.3
collecting ... /home/pearu/miniconda3/envs/pytorch-cuda-dev/lib/python3.9/site-packages/scipy/__init__.py:138: UserWarning: A NumPy version >=1.16.5 and <1.23.0 is required for this version of SciPy (detected version 1.26.0)
  warnings.warn(f"A NumPy version >={np_minversion} and <{np_maxversion} is required for this version of "
collected 6903 items / 6889 deselected / 14 selected                                                                                                                                                       
Running 14 items in this shard

test/inductor/test_torchinductor_opinfo.py::TestInductorOpInfoCPU::test_comprehensive_put_cpu_bool FAILED [0.1232s]

================================================================================================= FAILURES =================================================================================================
__________________________________________________________________________ TestInductorOpInfoCPU.test_comprehensive_put_cpu_bool ___________________________________________________________________________
Traceback (most recent call last):
  File "/home/pearu/git/pytorch/pytorch-csc/test/inductor/test_torchinductor_opinfo.py", line 654, in test_comprehensive
    raise e
  File "/home/pearu/git/pytorch/pytorch-csc/test/inductor/test_torchinductor_opinfo.py", line 621, in test_comprehensive
    no_python, has_rng_op = do_nopython_and_has_rng(fn, args, kwargs)
  File "/home/pearu/git/pytorch/pytorch-csc/test/inductor/test_torchinductor_opinfo.py", line 550, in do_nopython_and_has_rng
    fn(*args, **kwargs)
  File "/home/pearu/git/pytorch/pytorch-csc/test/inductor/test_torchinductor_opinfo.py", line 506, in fn
    return func(*args, **kwargs)
  File "/home/pearu/git/pytorch/pytorch-csc/test/inductor/test_torchinductor_opinfo.py", line 535, in __torch_dispatch__
    return func(*args, **kwargs)
  File "/home/pearu/git/pytorch/pytorch-csc/torch/_ops.py", line 571, in __call__
    return self_._op(*args, **kwargs)
  File "/home/pearu/git/pytorch/pytorch-csc/torch/utils/_stats.py", line 20, in wrapper
    return fn(*args, **kwargs)
  File "/home/pearu/git/pytorch/pytorch-csc/torch/_subclasses/fake_tensor.py", line 880, in __torch_dispatch__
    return self.dispatch(func, types, args, kwargs)
  File "/home/pearu/git/pytorch/pytorch-csc/torch/_subclasses/fake_tensor.py", line 1219, in dispatch
    return self._cached_dispatch_impl(func, types, args, kwargs)
  File "/home/pearu/git/pytorch/pytorch-csc/torch/_subclasses/fake_tensor.py", line 950, in _cached_dispatch_impl
    output = self._dispatch_impl(func, types, args, kwargs)
  File "/home/pearu/git/pytorch/pytorch-csc/torch/_subclasses/fake_tensor.py", line 1436, in _dispatch_impl
    r = func(*args, **kwargs)
  File "/home/pearu/git/pytorch/pytorch-csc/torch/_ops.py", line 571, in __call__
    return self_._op(*args, **kwargs)
  File "/home/pearu/git/pytorch/pytorch-csc/torch/_decomp/decompositions.py", line 4380, in inplace_op
    out = outplace_op(*args, **kwargs)
  File "/home/pearu/git/pytorch/pytorch-csc/torch/_ops.py", line 825, in __call__
    return self_._op(*args, **(kwargs or {}))
<snip>
  File "/home/pearu/git/pytorch/pytorch-csc/torch/_ops.py", line 825, in __call__
    return self_._op(*args, **(kwargs or {}))
  File "/home/pearu/git/pytorch/pytorch-csc/torch/_decomp/decompositions.py", line 4380, in inplace_op
    out = outplace_op(*args, **kwargs)
  File "/home/pearu/git/pytorch/pytorch-csc/torch/_ops.py", line 825, in __call__
    return self_._op(*args, **(kwargs or {}))
  File "/home/pearu/git/pytorch/pytorch-csc/torch/_prims_common/wrappers.py", line 252, in _fn
    result = fn(*args, **kwargs)
  File "/home/pearu/git/pytorch/pytorch-csc/torch/_refs/__init__.py", line 1927, in clone
    result = prims.clone(a, memory_format=memory_format)
  File "/home/pearu/git/pytorch/pytorch-csc/torch/_ops.py", line 571, in __call__
    return self_._op(*args, **kwargs)
  File "/home/pearu/git/pytorch/pytorch-csc/torch/_prims/__init__.py", line 651, in _clone_meta
    strides = utils.compute_elementwise_output_strides(input)
  File "/home/pearu/git/pytorch/pytorch-csc/torch/_prims_common/__init__.py", line 514, in compute_elementwise_output_strides
    new_strides = make_contiguous_strides_for(permuted_shape)
  File "/home/pearu/git/pytorch/pytorch-csc/torch/_prims_common/__init__.py", line 1565, in make_contiguous_strides_for
    validate_shape(shape)
  File "/home/pearu/git/pytorch/pytorch-csc/torch/_prims_common/__init__.py", line 562, in validate_shape
    assert isinstance(shape, Sequence), type(shape)
  File "/home/pearu/miniconda3/envs/pytorch-cuda-dev/lib/python3.9/typing.py", line 697, in __instancecheck__
    return self.__subclasscheck__(type(obj))
  File "/home/pearu/miniconda3/envs/pytorch-cuda-dev/lib/python3.9/typing.py", line 829, in __subclasscheck__
    return issubclass(cls, self.__origin__)
  File "/home/pearu/miniconda3/envs/pytorch-cuda-dev/lib/python3.9/abc.py", line 102, in __subclasscheck__
    return _abc_subclasscheck(cls, subclass)
RecursionError: maximum recursion depth exceeded in comparison
```

<details>

pearu committed:
Update on "Add a decomposition for torch.put, 2."

As in the title. It is an updated copy of https://github.com/pytorch/pytorch/pull/115306 .




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames

[ghstack-poisoned]

lezcano commented on torch/_inductor/decomposition.py:
```diff
@@ -647,3 +647,19 @@ def masked_scatter(self, mask, source):
+    return flattened.reshape(self.shape)
+
+
+@register_decomposition(aten.put_)
+def put_(self, index, source, accumulate=False):
+    out = aten.put(self, index, source, accumulate=accumulate)
+    return self.copy_(out)
```
Not needed anymore

pearu replied:
Not sure about this, see https://github.com/pytorch/pytorch/pull/120179#discussion_r1494800076

lezcano commented on torch/_inductor/decomposition.py:
```diff
@@ -647,3 +647,19 @@ def masked_scatter(self, mask, source):
+
+@register_decomposition(aten.put)
+def put(self, index, source, accumulate=False):
+    stride = utils.compute_elementwise_output_strides(self)
```
unused

pearu replied:
removed

lezcano commented on torch/_inductor/codegen/cpp.py:
```diff
@@ -1547,6 +1547,8 @@ def store(self, name, index, value, mode=None):
             if not config.cpp.dynamic_threads and self.num_threads == 1:
                 line = f"{var}[{cexpr_index(index)}] += {value};"
             else:
+                if value.dtype in [torch.float16]:
```
Shouldn't this be:
```suggestion
                if V.graph.get_dtype(name) == torch.float16:
```
?

lezcano replied:
and leave a comment, given that if anyone updates the behaviour in the load, we should update this.

lezcano replied:
cc @jgong5

pearu replied:
Now using unconditional casting.

pearu committed:
Update on "Add a decomposition for torch.put, 2."

As in the title. It is an updated copy of https://github.com/pytorch/pytorch/pull/115306 .




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames

[ghstack-poisoned]

pearu committed:
Update on "Add a decomposition for torch.put, 2."

As in the title. It is an updated copy of https://github.com/pytorch/pytorch/pull/115306 .




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames

[ghstack-poisoned]

lezcano reviewed (approved):
Ok, so the issue with the inplace op is the following:
- `put` is implemented in eager mode as `self.clone().put_` (CompositeExplicit)
- When running this test we don't run it with inductor decomps (not at the test level, they happen to be executed later on)

Adding `put_` as inductor only lgtm.

The decomp changes lgtm.

@jgong5 can you have a look at the `cpp.py` change?

jgong5 commented on torch/_inductor/codegen/cpp.py:
```diff
@@ -1547,6 +1547,9 @@ def store(self, name, index, value, mode=None):
             else:
+                dtype = V.graph.get_dtype(name)
+                # mirroring static_cast<float>(...) in load:
+                value = f"static_cast<{DTYPE_TO_CPP[dtype]}>({value})"
```
Thanks, I missed the ping. The change LGTM.

pearu committed:
Update on "Add a decomposition for torch.put, 2."

As in the title. It is an updated copy of https://github.com/pytorch/pytorch/pull/115306 .




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames

[ghstack-poisoned]

pearu commented:
@pytorchbot merge -f "The 4 CI failures are not related to the changes in the PR"
